### ISSUE TITLE
Dulwich Git history contains invalid objects

### ISSUE THREAD

**Issue by jelmer** [jelmer/dulwich#620]
The Dulwich Git history contains a few ".git" directories in early commits, at a time when C git still allowed this.

from "git fsck --full":

warning in tree eaea53f8085ad0f9cd299b534774dfaef2bc26f1: hasDotgit: contains '.git'
warning in tree e5f1393aeea756d43d574504d932a887a36f1ce8: hasDotgit: contains '.git'
warning in tree de750728425824837d8e774361b1033cbc0f2843: hasDotgit: contains '.git'
warning in tree 4d73ba3255064784acec2a5102790c707d042043: hasDotgit: contains '.git'
warning in tree eaea53f8085ad0f9cd299b534774dfaef2bc26f1: hasDotgit: contains '.git'

github now refuses packs with trees like this.

Unfortunately the only way to fix this is to rewrite history :(

**Comment by rctay** [comment]
Ran into this today while doing a `git clone`, got around this by using `git clone --depth=1`. (Note to others: should you need further history, you can "deepen" your git's shallow clone with `git fetch --depth=100` etc.)

**Comment by nanonyme** [comment]
Yeah so the problem is GitHub and Gitlab and others in fact reject this kind of a repository now. I'm not sure there's another way to recover than archiving this repo with some other name and creating new repo with fully rewritten history.

**Comment by nanonyme** [comment]
This wouldn't be so bad if it was only branches but tags are probably broken as well

**Comment by nanonyme** [comment]
That is, it's not just clone that fails. Git push does as well

**Comment by jelmer** [comment]
@nanonyme 

Pushing tags to existing repositories (and Dulwich forks) is fine; the only thing that doesn't work is pushing the existing Dulwich history to an empty repository.

This isn't an actual problem for Dulwich development. @nanonyme where are you finding this is problematic?

**Comment by jelmer** [comment]
Is publishing a fork / shallow clone not an option?

**Comment by jelmer** [comment]
I've been hesitant to go the route of rewriting history, because it breaks all forks and PRs. However, if having these commits is too painful we can consider it.

**Comment by nanonyme** [comment]
I completely understand, it's nothing to take too lightly and the bar rises as project matures.

**Comment by jelmer** [comment]
If you wanted to work on it, I think here is a possible way to approach this:

1) Provide a script to rewrite history, and add a headsup in README about what we're doing
2) We can provide a revised branch in this repository for a while
3) Make sure we're happy with the rewritten history
4) Switch over

If we're going to do this, there is at least one other thing we should fix. IIRC there is a commit with a dropped digit for the timestamp somewhere in history, which means that it is reported as being decades old rather than from 2014 or so - we should probably fix that at the same time.

**Comment by jelmer** [comment]
There is now a devscripts/fix-history.py script which does this.

**Comment by shymega-ct** [comment]
Hi, we at [freedesktop-sdk](https://gitlab.com/freedesktop-sdk/freedesktop-sdk) do need tagged releases fixed as well.

When I run `git fsck` on the dulwich repo, I get the following:

The checkout has 6465a4e0d27a0902314fa64526c84134fac40450 at HEAD.

```shell
❯ git fsck
Checking ref database: 100% (1/1), done.
Checking object directories: 100% (256/256), done.
warning in tree 4e06937e7b09c4932a75d85a342fc45047c840bc: badFilemode: contains bad file modes
warning in tree dc6afd8bd24556659fcbe6438ef6fc9f9c45db7a: badFilemode: contains bad file modes
warning in tree 976c4e2799faa3d4e67bbedd8d1d3327d09c1b2d: badFilemode: contains bad file modes
warning in tree e5f1393aeea756d43d574504d932a887a36f1ce8: hasDotgit: contains '.git'
warning in tree de750728425824837d8e774361b1033cbc0f2843: hasDotgit: contains '.git'
warning in tree 4d73ba3255064784acec2a5102790c707d042043: hasDotgit: contains '.git'
warning in tree eaea53f8085ad0f9cd299b534774dfaef2bc26f1: hasDotgit: contains '.git'
Checking objects: 100% (39363/39363), done.
```

It would appear this is only fixed in the `main` branch, and needs to be fixed in tags too.

Would it be reasonable to address this in this issue, or should I open a new issue to discuss further?

**Comment by shymega-ct** [comment]
Interestingly enough, I used to be able to get output from `git describe`, but not after a fresh clone.

**Comment by jelmer** [comment]
Which tags are still broken? As far as I can tell, the only ref that still references the broken commits is the "oldmaster" branch - which is intentional.
